src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                                                                               bool trampoline,
+                                                                               CancellationToken cancellationToken)
+        {
+            var firstCallSite = callSites.Keys.First();
+            var currentSolution = firstCallSite.Project.Solution;
+            var modifiedSolution = currentSolution;
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+
+            foreach (var grouping in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var project = grouping.Key;
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var keyValuePair in grouping)
+                {
+                    var document = keyValuePair.Key;
+                    var invocationExpressionList = keyValuePair.Value;
+                    var generator = SyntaxGenerator.GetGenerator(document);
+                    var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+                    var invocationSemanticModel = compilation.GetSemanticModel(await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false));
+
+                    var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var editor = new SyntaxEditor(root, generator);
+                    var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                    var oldMethodDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => IsMethodDeclaration(node), ascendOutOfTrivia: true)!;
+                    var insertionIndex = GetInsertionIndex(semanticDocument, compilation, syntaxFacts, oldMethodDeclaration, cancellationToken);
+
+                    if (trampoline)
+                    {
+                        foreach (var invocationExpression in invocationExpressionList)
+                        {
+                            editor.ReplaceNode(invocationExpression, (currentInvocation, _) =>
+                            {
+                                var invocationArguments = syntaxFacts.GetArgumentsOfInvocationExpression(currentInvocation);
+                                var methodName = generator.IdentifierName(newMethodIdentifier);
+                                var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+                                var newMethodInvocation = generator.InvocationExpression(methodName, invocationArguments);
+                                var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+                                return editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
```

Comment:
please extract thsi to a helper, and please doc what it is doing.

Reply:
extracted